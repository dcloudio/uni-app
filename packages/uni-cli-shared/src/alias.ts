import { initModuleAlias } from './hbx/alias'

initModuleAlias()
