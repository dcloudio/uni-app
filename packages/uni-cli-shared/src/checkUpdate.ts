import fs from 'fs-extra'
import os from 'os'
import path from 'path'
import debug from 'debug'
import crypto from 'crypto'
import { request } from 'https'
import compareVersions from 'compare-versions'
import { hasOwn, isString, isPlainObject } from '@vue/shared'

import { parseManifestJsonOnce } from './json'
import { isInHBuilderX } from './hbx'

const debugCheckUpdate = debug('uni:check-update')

interface CheckUpdateOptions {
  inputDir: string
  compilerVersion: string
  versionType: 'a' | 'r' // alpha | release
}

interface CheckUpdatePlatform {
  appid?: string
  dev: number
  build: number
}

interface CheckUpdateCache {
  vid: string
  lastCheck: number
  newVersion?: string
  note?: string
  [name: string]: CheckUpdatePlatform | undefined | string | number
}

interface CheckVersionRequest {
  vv: 3
  device: string
  appid?: string
  vid?: string
  vtype: CheckUpdateOptions['versionType']
  vcode: string
  [name: string]: CheckUpdatePlatform | undefined | string | number
}

interface CheckVersionResponse {
  code: number
  msg: string
  isUpdate: boolean
  newVersion: string
  note: string
}

const INTERVAL = 1000 * 60 * 60 * 24

export async function checkUpdate(options: CheckUpdateOptions) {
  if (process.env.CI) {
    debugCheckUpdate('isInCI')
    return
  }
  if (isInHBuilderX()) {
    debugCheckUpdate('isInHBuilderX')
    return
  }
  const { inputDir, compilerVersion } = options
  const updateCache = readCheckUpdateCache(inputDir)
  debugCheckUpdate('read.cache', updateCache)
  const res = checkLocalCache(updateCache, compilerVersion)
  if (res) {
    if (isString(res)) {
      console.log()
      console.log(res)
    }
  } else {
    await checkVersion(
      options,
      normalizeUpdateCache(updateCache, parseManifestJsonOnce(inputDir))
    )
  }
  writeCheckUpdateCache(
    inputDir,
    statUpdateCache(normalizeUpdateCache(updateCache))
  )
}

function normalizeUpdateCache(
  updateCache: CheckUpdateCache,
  manifestJson?: Record<string, any>
) {
  const platform = process.env.UNI_PLATFORM
  if (!updateCache[platform]) {
    updateCache[platform] = {
      appid: '',
      dev: 0,
      build: 0,
    }
  }
  if (manifestJson) {
    const platformOptions =
      manifestJson[platform === 'app' ? 'app-plus' : platform]
    ;(updateCache[platform] as CheckUpdatePlatform).appid = platformOptions
      ? platformOptions.appid || platformOptions.package || ''
      : ''
  }
  return updateCache
}

function statUpdateCache(updateCache: CheckUpdateCache) {
  debugCheckUpdate('stat.before', updateCache)
  const platform = process.env.UNI_PLATFORM
  const type = process.env.NODE_ENV === 'production' ? 'build' : 'dev'
  const platformOptions = updateCache[platform] as CheckUpdatePlatform
  platformOptions[type] = (platformOptions[type] || 0) + 1
  debugCheckUpdate('stat.after', updateCache)
  return updateCache
}

function getFilepath(inputDir: string, filename: string) {
  return path.resolve(os.tmpdir(), 'uni-app-cli', md5(inputDir), filename)
}

function getCheckUpdateFilepath(inputDir: string) {
  return getFilepath(inputDir, 'check-update.json')
}

function generateVid() {
  let result = ''
  for (let i = 0; i < 4; i++) {
    result += ((65536 * (1 + Math.random())) | 0).toString(16).substring(1)
  }
  return 'UNI_' + result.toUpperCase()
}

function createCheckUpdateCache(vid: string = generateVid()): CheckUpdateCache {
  return {
    vid: generateVid(),
    lastCheck: 0,
  }
}

function readCheckUpdateCache(inputDir: string) {
  const updateFilepath = getCheckUpdateFilepath(inputDir)
  debugCheckUpdate('read:', updateFilepath)
  if (fs.existsSync(updateFilepath)) {
    try {
      return require(updateFilepath) as CheckUpdateCache
    } catch (e) {
      debugCheckUpdate('read.error', e)
    }
  }
  return createCheckUpdateCache()
}

/**
 * 检查本地缓存，返回 false 表示需要执行云端检查，返回 true 表示，无需云端检查，返回 string 表示，无需云端检测，且有更新
 * @param inputDir
 * @param compilerVersion
 * @param interval
 * @returns
 */
export function checkLocalCache(
  updateCache: CheckUpdateCache,
  compilerVersion: string,
  interval: number = INTERVAL
) {
  if (!updateCache.lastCheck) {
    debugCheckUpdate('cache: lastCheck not found')
    return false
  }
  if (Date.now() - updateCache.lastCheck > interval) {
    debugCheckUpdate('cache: lastCheck > interval')
    return false
  }
  if (
    updateCache.newVersion &&
    compareVersions(updateCache.newVersion, compilerVersion) > 0
  ) {
    debugCheckUpdate('cache: find new version')
    return updateCache.note!
  }
  return true
}

function writeCheckUpdateCache(
  inputDir: string,
  updateCache: CheckUpdateCache
) {
  const filepath = getCheckUpdateFilepath(inputDir)
  debugCheckUpdate('write:', filepath, updateCache)
  try {
    fs.outputFileSync(filepath, JSON.stringify(updateCache))
  } catch (e) {
    debugCheckUpdate('write.error', e)
  }
}

export function md5(str: string) {
  return crypto.createHash('md5').update(str).digest('hex')
}

export function getMac() {
  let mac = ''
  const network = os.networkInterfaces()
  for (const key in network) {
    const array = network[key]!
    for (let i = 0; i < array.length; i++) {
      const item = array[i]
      if (!item.family || (item.mac && item.mac === '00:00:00:00:00:00')) {
        continue
      }

      if (
        // Node < v18
        isString(item.family) &&
        (item.family === 'IPv4' || item.family === 'IPv6')
      ) {
        mac = item.mac
        break
      } else if (
        // Node >= v18
        typeof item.family === 'number' &&
        (item.family === 4 || item.family === 6)
      ) {
        mac = (item as any).mac
        break
      }
    }
  }
  return mac
}

export function createPostData(
  { versionType, compilerVersion }: CheckUpdateOptions,
  manifestJson: Record<string, any>,
  updateCache: CheckUpdateCache
) {
  const data: CheckVersionRequest = {
    vv: 3,
    device: md5(getMac()),
    vtype: versionType,
    vcode: compilerVersion,
  }
  if (manifestJson.appid) {
    data.appid = manifestJson.appid
  } else {
    data.vid = updateCache.vid
  }
  Object.keys(updateCache).forEach((name) => {
    const value = updateCache[name]
    if (
      isPlainObject(value) &&
      (hasOwn(value, 'dev') || hasOwn(value, 'build'))
    ) {
      data[name] = value
    }
  })
  return JSON.stringify(data)
}

function handleCheckVersion(
  { code, isUpdate, newVersion, note }: CheckVersionResponse,
  updateCache: CheckUpdateCache
) {
  if (code !== 0) {
    return
  }
  // clear
  Object.keys(updateCache).forEach((key) => {
    if (key !== 'vid') delete updateCache[key]
  })
  updateCache.lastCheck = Date.now()
  if (isUpdate) {
    updateCache.note = note
    updateCache.newVersion = newVersion
  } else {
    delete updateCache.note
    delete updateCache.newVersion
  }
}

const HOSTNAME = 'uniapp.dcloud.net.cn'
const PATH = '/update/cli'

function checkVersion(
  options: CheckUpdateOptions,
  updateCache: CheckUpdateCache
) {
  return new Promise((resolve) => {
    const postData = JSON.stringify({
      id: createPostData(
        options,
        parseManifestJsonOnce(options.inputDir),
        updateCache
      ),
    })
    let responseData = ''
    const req = request(
      {
        hostname: HOSTNAME,
        path: PATH,
        port: 443,
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': postData.length,
        },
      },
      (res) => {
        res.setEncoding('utf8')
        res.on('data', (chunk) => {
          responseData += chunk
        })
        res.on('end', () => {
          debugCheckUpdate('response: ', responseData)
          try {
            handleCheckVersion(JSON.parse(responseData), updateCache)
          } catch (e) {}
          resolve(true)
        })
        res.on('error', (e) => {
          debugCheckUpdate('response.error:', e)
          resolve(false)
        })
      }
    ).on('error', (e) => {
      debugCheckUpdate('request.error:', e)
      resolve(false)
    })
    debugCheckUpdate('request: ', postData)
    req.write(postData)
    req.end()
  })
}
