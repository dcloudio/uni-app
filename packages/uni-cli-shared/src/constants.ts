export const PUBLIC_DIR = 'static'
export const EXTNAME_JS = ['.js', '.ts', '.jsx', '.tsx']
export const X_EXTNAME_JS = ['.uts', ...EXTNAME_JS]
export const EXTNAME_TS = ['.ts', '.tsx']
export const EXTNAME_VUE = ['.vue', '.nvue', '.uvue']
export const X_EXTNAME_VUE = ['.uvue', '.vue']
export const EXTNAME_VUE_TEMPLATE = ['.vue', '.nvue', '.uvue', '.jsx', '.tsx']
export const EXTNAME_VUE_RE = /\.(vue|nvue|uvue)$/
export const EXTNAME_JS_RE = /\.(js|jsx|ts|uts|tsx|mjs)$/
export const EXTNAME_TS_RE = /\.tsx?$/

const COMMON_EXTENSIONS = [
  '.uts',
  '.mjs',
  '.js',
  '.ts',
  '.jsx',
  '.tsx',
  '.json',
]
export const extensions = COMMON_EXTENSIONS.concat(EXTNAME_VUE)

export const uni_app_x_extensions = COMMON_EXTENSIONS.concat(['.uvue', '.vue'])

export const PAGES_JSON_JS = 'pages-json-js'
export const PAGES_JSON_UTS = 'pages-json-uts'
export const MANIFEST_JSON_JS = 'manifest-json-js'
export const MANIFEST_JSON_UTS = 'manifest-json-uts'

export const JSON_JS_MAP = {
  'pages.json': PAGES_JSON_JS,
  'manifest.json': MANIFEST_JSON_JS,
} as const
export const ASSETS_INLINE_LIMIT = 40 * 1024

export const APP_SERVICE_FILENAME = 'app-service.js'
export const APP_CONFIG = 'app-config.js'
export const APP_CONFIG_SERVICE = 'app-config-service.js'

export const BINDING_COMPONENTS = '__BINDING_COMPONENTS__'
// APP 平台解析页面后缀的优先级
export const PAGE_EXTNAME_APP = ['.nvue', '.vue', '.tsx', '.jsx', '.js']
// 其他平台解析页面后缀的优先级
export const PAGE_EXTNAME = ['.vue', '.nvue', '.tsx', '.jsx', '.js']

export const X_PAGE_EXTNAME = ['.uvue', '.vue', '.tsx', '.jsx', '.js']
export const X_PAGE_EXTNAME_APP = ['.uvue', '.tsx', '.jsx', '.js']

export const H5_API_STYLE_PATH = '@dcloudio/uni-h5/style/api/'
export const H5_FRAMEWORK_STYLE_PATH = '@dcloudio/uni-h5/style/framework/'
export const H5_COMPONENTS_STYLE_PATH = '@dcloudio/uni-h5/style/'
export const BASE_COMPONENTS_STYLE_PATH = '@dcloudio/uni-components/style/'
export const X_BASE_COMPONENTS_STYLE_PATH = '@dcloudio/uni-components/style-x/'

export const COMMON_EXCLUDE = [
  /\/pages\.json\.js$/,
  /\/manifest\.json\.js$/,
  /\/vite\//,
  /\/@vue\//,
  /\/vue-router\//,
  /\/vuex\//,
  /\/vue-i18n\//,
  /\/@dcloudio\/uni-h5-vue/,
  /\/@dcloudio\/uni-shared/,
]

export const KNOWN_ASSET_TYPES = [
  // images
  'png',
  'jpe?g',
  'gif',
  'svg',
  'ico',
  'webp',
  'avif',

  // media
  'mp4',
  'webm',
  'ogg',
  'mp3',
  'wav',
  'flac',
  'aac',

  // fonts
  'woff2?',
  'eot',
  'ttf',
  'otf',

  // other
  'pdf',
  'txt',
]

export const DEFAULT_ASSETS_RE = new RegExp(
  `\\.(` + KNOWN_ASSET_TYPES.join('|') + `)(\\?.*)?$`
)

export const TEXT_STYLE = ['black', 'white']
