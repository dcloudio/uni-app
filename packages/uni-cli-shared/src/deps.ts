import {
  BASE_COMPONENTS_STYLE_PATH,
  H5_API_STYLE_PATH,
  H5_COMPONENTS_STYLE_PATH,
} from './constants'

const RESIZE_SENSOR_CSS = BASE_COMPONENTS_STYLE_PATH + 'resize-sensor.css'
const REFRESHER_CSS = BASE_COMPONENTS_STYLE_PATH + 'refresher.css'

export const API_DEPS_CSS = (isX: boolean) => {
  return {
    showModal: [`${H5_API_STYLE_PATH}modal.css`],
    showToast: [`${H5_API_STYLE_PATH}toast.css`],
    showActionSheet: [`${H5_API_STYLE_PATH}action-sheet.css`],
    previewImage: [
      RESIZE_SENSOR_CSS,
      `${BASE_COMPONENTS_STYLE_PATH}swiper.css`,
      `${BASE_COMPONENTS_STYLE_PATH}swiper-item.css`,
      `${BASE_COMPONENTS_STYLE_PATH}movable-area.css`,
      `${BASE_COMPONENTS_STYLE_PATH}movable-view.css`,
    ],
    openLocation: [`${H5_API_STYLE_PATH}location-view.css`],
    chooseLocation: [
      ...(isX
        ? [
            `${BASE_COMPONENTS_STYLE_PATH}/view.css`,
            `${BASE_COMPONENTS_STYLE_PATH}/text.css`,
          ]
        : [`${H5_API_STYLE_PATH}/location-picker.css`]),
      `${BASE_COMPONENTS_STYLE_PATH}/input.css`,
      `${H5_COMPONENTS_STYLE_PATH}/map.css`,
      `${BASE_COMPONENTS_STYLE_PATH}/scroll-view.css`,
    ],
  }
}
export const COMPONENT_DEPS_CSS = (isX: boolean) => {
  return {
    canvas: [RESIZE_SENSOR_CSS],
    image: [RESIZE_SENSOR_CSS],
    'movable-area': [RESIZE_SENSOR_CSS],
    'picker-view': [RESIZE_SENSOR_CSS],
    'picker-view-column': [RESIZE_SENSOR_CSS],
    'rich-text': [RESIZE_SENSOR_CSS],
    textarea: [RESIZE_SENSOR_CSS],
    'web-view': [RESIZE_SENSOR_CSS],
    picker: [
      RESIZE_SENSOR_CSS,
      `${BASE_COMPONENTS_STYLE_PATH}picker-view.css`,
      `${BASE_COMPONENTS_STYLE_PATH}picker-view-column.css`,
    ],
    'scroll-view': [REFRESHER_CSS],
    'list-view': [RESIZE_SENSOR_CSS, REFRESHER_CSS],
  }
}
