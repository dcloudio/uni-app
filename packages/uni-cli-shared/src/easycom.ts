import fs from 'fs'
import path from 'path'
import debug from 'debug'

import { camelize, capitalize, extend } from '@vue/shared'
import { createFilter } from '@rollup/pluginutils'

import { once } from '@dcloudio/uni-shared'
import { normalizePath } from './utils'
import { parsePagesJson, parsePagesJsonOnce } from './json/pages'
import { M } from './messages'
import {
  clearUTSComponents,
  clearUTSCustomElements,
  initUTSComponents,
  initUTSCustomElements,
} from './uts'
import { genUTSClassName } from './utsUtils'

interface EasycomOption {
  isX?: boolean
  dirs?: string[]
  rootDir: string
  extensions: string[]
  autoscan?: boolean
  custom?: EasycomCustom
}
export interface EasycomMatcher {
  name: string
  pattern: RegExp
  replacement: string
}
interface EasycomCustom {
  [key: string]: string
}

const debugEasycom = debug('uni:easycom')

const easycoms: EasycomMatcher[] = []

const easycomsCache = new Map<string, string>()
const easycomsInvalidCache = new Set<string>()

let hasEasycom = false

function clearEasycom() {
  easycoms.length = 0

  easycomsCache.clear()
  easycomsInvalidCache.clear()
}

export function initEasycoms(
  inputDir: string,
  {
    dirs,
    platform,
    isX,
  }: { dirs: string[]; platform: UniApp.PLATFORM; isX?: boolean }
) {
  const componentsDir = path.resolve(inputDir, 'components')
  const uniModulesDir = path.resolve(inputDir, 'uni_modules')
  const initEasycomOptions = (pagesJson?: UniApp.PagesJson) => {
    // 初始化时，从once中读取缓存，refresh时，实时读取
    const { easycom } = pagesJson || parsePagesJson(inputDir, platform, false)
    const easycomOptions: EasycomOption = {
      isX,
      dirs:
        easycom && easycom.autoscan === false
          ? [...dirs] // 禁止自动扫描
          : [
              ...dirs,
              componentsDir,
              ...initUniModulesEasycomDirs(uniModulesDir),
            ],
      rootDir: inputDir,
      autoscan: !!(easycom && easycom.autoscan),
      custom: (easycom && easycom.custom) || {},
      extensions: [...(isX ? ['.uvue'] : []), ...['.vue', '.jsx', '.tsx']],
    }
    debugEasycom(easycomOptions)
    return easycomOptions
  }

  const easyComOptions = initEasycomOptions(
    parsePagesJsonOnce(inputDir, platform)
  )

  const initUTSEasycom = () => {
    initUTSComponents(inputDir, platform).forEach((item) => {
      const index = easycoms.findIndex((easycom) => item.name === easycom.name)
      if (index > -1) {
        easycoms.splice(index, 1, item)
      } else {
        easycoms.push(item)
      }
    })
    if (isX && (globalThis as any).uts2jsSourceCodeMap) {
      ;(globalThis as any).uts2jsSourceCodeMap.initUts2jsEasycom(easycoms)
    }
  }

  const initUTSEasycomCustomElements = () => {
    initUTSCustomElements(inputDir, platform).forEach((item) => {
      const index = easycoms.findIndex((easycom) => item.name === easycom.name)
      if (index > -1) {
        easycoms.splice(index, 1, item)
      } else {
        easycoms.push(item)
      }
    })
  }

  // ext-api 模式下，不存在 easycom 特性
  if (process.env.UNI_COMPILE_TARGET !== 'ext-api') {
    clearEasycom()
    clearUTSComponents()
    clearUTSCustomElements()
    initEasycom(easyComOptions)
    initUTSEasycomCustomElements()
    initUTSEasycom()
  }
  const componentExtNames = isX ? 'uvue|vue' : 'vue'
  const res = {
    easyComOptions,
    filter: createFilter(
      [
        'components/*/*.(' + componentExtNames + '|jsx|tsx)',
        'uni_modules/*/components/*/*.(' + componentExtNames + '|jsx|tsx)',
        'utssdk/*/**/*.(' + componentExtNames + ')',
        'uni_modules/*/utssdk/*/*.(' + componentExtNames + ')',
      ],
      [],
      {
        resolve: inputDir,
      }
    ),
    refresh() {
      res.easyComOptions = initEasycomOptions()
      if (process.env.UNI_COMPILE_TARGET !== 'ext-api') {
        clearEasycom()
        clearUTSComponents()
        clearUTSCustomElements()
        initEasycom(easyComOptions)
        initUTSEasycomCustomElements()
        initUTSEasycom()
      }
    },
    easycoms,
  }
  return res
}

export const initEasycomsOnce = once(initEasycoms)

function initUniModulesEasycomDirs(
  uniModulesDir: string,
  componentsDir: string = 'components'
) {
  if (!fs.existsSync(uniModulesDir)) {
    return []
  }
  return fs
    .readdirSync(uniModulesDir)
    .map((uniModuleDir) => {
      const uniModuleComponentsDir = path.resolve(
        uniModulesDir,
        uniModuleDir,
        componentsDir
      )
      if (fs.existsSync(uniModuleComponentsDir)) {
        return uniModuleComponentsDir
      }
    })
    .filter<string>(Boolean as any)
}

function initEasycom({
  isX,
  dirs,
  rootDir,
  custom,
  extensions,
}: EasycomOption) {
  rootDir = normalizePath(rootDir)
  const easycomsObj = Object.create(null)
  if (dirs && dirs.length && rootDir) {
    const autoEasyComObj = initAutoScanEasycoms(dirs, rootDir, extensions)
    if (isX) {
      Object.keys(autoEasyComObj).forEach((tagName) => {
        let source = autoEasyComObj[tagName]
        tagName = tagName.slice(1, -1)
        if (path.isAbsolute(source) && source.startsWith(rootDir)) {
          source = '@/' + normalizePath(path.relative(rootDir, source))
        }
        // 加密插件easycom类型导入
        if (source.includes('?uts-proxy')) {
          const moduleId = path.basename(source.split('?uts-proxy')[0])
          source = `uts.sdk.modules.${camelize(moduleId)}`
        }
        const ident = genUTSComponentPublicInstanceIdent(tagName)
        addUTSEasyComAutoImports(source, [ident, ident])
      })
    }
    extend(easycomsObj, autoEasyComObj)
  }
  if (custom) {
    Object.keys(custom).forEach((name) => {
      const componentPath = custom[name]
      easycomsObj[name] = componentPath.startsWith('@/')
        ? normalizePath(path.join(rootDir!, componentPath.slice(2)))
        : componentPath
    })
  }
  Object.keys(easycomsObj).forEach((name) => {
    easycoms.push({
      name:
        name.startsWith('^') && name.endsWith('$') ? name.slice(1, -1) : name,
      pattern: new RegExp(name),
      replacement: easycomsObj[name],
    })
  })
  debugEasycom(easycoms)
  hasEasycom = !!easycoms.length
  return easycoms
}

export function matchEasycom(tag: string) {
  if (!hasEasycom) {
    return
  }
  let source = easycomsCache.get(tag)
  if (source) {
    return source
  }
  if (easycomsInvalidCache.has(tag)) {
    return false
  }
  const matcher = easycoms.find((matcher) => matcher.pattern.test(tag))
  if (!matcher) {
    easycomsInvalidCache.add(tag)
    return false
  }
  source = tag.replace(matcher.pattern, matcher.replacement)
  easycomsCache.set(tag, source)
  debugEasycom('matchEasycom', tag, source)
  return source
}

const isDir = (path: string) => {
  const stat = fs.lstatSync(path)
  if (stat.isDirectory()) {
    return true
  } else if (stat.isSymbolicLink()) {
    return fs.lstatSync(fs.realpathSync(path)).isDirectory()
  }
  return false
}

function initAutoScanEasycom(
  dir: string,
  rootDir: string,
  extensions: string[]
): Record<string, string> {
  if (!path.isAbsolute(dir)) {
    dir = path.resolve(rootDir, dir)
  }
  const easycoms = Object.create(null)
  if (!fs.existsSync(dir)) {
    return easycoms
  }
  const is_uni_modules =
    path.basename(path.resolve(dir, '../..')) === 'uni_modules'
  const is_easycom_encrypt_uni_modules = // uni_modules模式不需要此逻辑
    process.env.UNI_COMPILE_TARGET !== 'uni_modules' &&
    is_uni_modules &&
    // 前端加密插件，不能包含utssdk目录
    fs.existsSync(path.resolve(dir, '../encrypt')) &&
    !fs.existsSync(path.resolve(dir, '../utssdk'))
  const uni_modules_plugin_id =
    is_easycom_encrypt_uni_modules && path.basename(path.resolve(dir, '..'))
  fs.readdirSync(dir).forEach((name) => {
    const folder = path.resolve(dir, name)
    if (!isDir(folder)) {
      return
    }
    const importDir = normalizePath(folder)
    const files = fs.readdirSync(folder)
    // 读取文件夹文件列表，比对文件名（fs.existsSync在大小写不敏感的系统会匹配不准确）
    for (let i = 0; i < extensions.length; i++) {
      const ext = extensions[i]
      if (files.includes(name + ext)) {
        easycoms[`^${name}$`] = is_easycom_encrypt_uni_modules
          ? normalizePath(
              path.join(
                rootDir,
                `uni_modules/${uni_modules_plugin_id}?${
                  // android 走 proxy
                  process.env.UNI_APP_X === 'true' &&
                  process.env.UNI_UTS_PLATFORM === 'app-android'
                    ? 'uts-proxy'
                    : 'uni_helpers'
                }`
              )
            )
          : `${importDir}/${name}${ext}`
        break
      }
    }
  })
  return easycoms
}

function initAutoScanEasycoms(
  dirs: string[],
  rootDir: string,
  extensions: string[]
) {
  const conflict: Record<string, string[]> = {}
  const res = dirs.reduce<Record<string, string>>(
    (easycoms: Record<string, string>, dir: string) => {
      const curEasycoms = initAutoScanEasycom(dir, rootDir, extensions)
      Object.keys(curEasycoms).forEach((name) => {
        // Use the first component when name conflict
        const componentPath = easycoms[name]
        if (!componentPath) {
          easycoms[name] = curEasycoms[name]
        } else {
          ;(conflict[componentPath] || (conflict[componentPath] = [])).push(
            normalizeComponentPath(curEasycoms[name], rootDir)
          )
        }
      })
      return easycoms
    },
    Object.create(null)
  )
  const conflictComponents = Object.keys(conflict)
  if (conflictComponents.length) {
    console.warn(M['easycom.conflict'])
    conflictComponents.forEach((com) => {
      console.warn(
        [normalizeComponentPath(com, rootDir), conflict[com]].join(',')
      )
    })
  }
  return res
}

function normalizeComponentPath(componentPath: string, rootDir: string) {
  return normalizePath(path.relative(rootDir, componentPath))
}

export function addImportDeclaration(
  importDeclarations: string[],
  local: string,
  source: string,
  imported?: string
) {
  importDeclarations.push(createImportDeclaration(local, source, imported))
  return local
}

function createImportDeclaration(
  local: string,
  source: string,
  imported?: string
) {
  if (imported) {
    return `import { ${imported} as ${local} } from '${source}';`
  }
  return `import ${local} from '${source}';`
}

const RESOLVE_EASYCOM_IMPORT_CODE = `import { resolveDynamicComponent as __resolveDynamicComponent } from 'vue';import { resolveEasycom } from '@dcloudio/uni-app';`

export function genResolveEasycomCode(
  importDeclarations: string[],
  code: string,
  name: string
) {
  if (!importDeclarations.includes(RESOLVE_EASYCOM_IMPORT_CODE)) {
    importDeclarations.push(RESOLVE_EASYCOM_IMPORT_CODE)
  }
  return `resolveEasycom(${code.replace(
    '_resolveComponent',
    '__resolveDynamicComponent'
  )}, ${name})`
}

export const UNI_EASYCOM_EXCLUDE = [/@dcloudio\/uni-h5/]

const utsEasyComAutoImports: Record<string, [[string, string]]> = {}

export function getUTSEasyComAutoImports() {
  return utsEasyComAutoImports
}

export function addUTSEasyComAutoImports(
  source: string,
  imports: [string, string]
) {
  if (!utsEasyComAutoImports[source]) {
    utsEasyComAutoImports[source] = [imports]
  } else {
    if (!utsEasyComAutoImports[source].find((item) => item[0] === imports[0])) {
      utsEasyComAutoImports[source].push(imports)
    }
  }
}

export function genUTSComponentPublicInstanceIdent(tagName: string) {
  return capitalize(camelize(tagName)) + 'ComponentPublicInstance'
}

export function genUTSComponentPublicInstanceImported(
  root: string,
  fileName: string
) {
  root = normalizePath(root)
  if (path.isAbsolute(fileName) && fileName.startsWith(root)) {
    fileName = normalizePath(path.relative(root, fileName))
  }
  if (fileName.startsWith('@/')) {
    return (
      genUTSClassName(fileName.replace('@/', '')) + 'ComponentPublicInstance'
    )
  }
  return genUTSClassName(fileName) + 'ComponentPublicInstance'
}
