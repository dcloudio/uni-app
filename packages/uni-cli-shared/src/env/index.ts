export { initDefine } from './define'
export { initAppProvide } from './provide'
