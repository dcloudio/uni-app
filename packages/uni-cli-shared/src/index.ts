export * from './fs'
export * from './mp'
export * from './url'
export * from './env'
export * from './hbx'
export * from './ssr'
export * from './vue'
export * from './uts'
export * from './logs'
export * from './i18n'
export * from './deps'
export * from './json'
export * from './vite'
export * from './utils'
export * from './easycom'
export * from './constants'
export * from './preprocess'
export * from './postcss'
export * from './filter'
export * from './esbuild'
export * from './resolve'
export * from './scripts'
export * from './platform'
export * from './utsUtils'

export {
  parseUniExtApi,
  parseUniExtApis,
  parseInjects,
  Define,
  DefineOptions,
  Defines,
  getUniExtApiProviderRegisters,
  formatExtApiProviderName,
} from './uni_modules'
export {
  parseUniModulesArtifacts,
  resolveEncryptUniModule,
} from './uni_modules.cloud'
import type { EncryptPackageJson } from './uni_modules.cloud'
export type EncryptArtifacts = EncryptPackageJson['uni_modules']['artifacts']

export { M } from './messages'

export * from './exports'
export { checkUpdate } from './checkUpdate'
