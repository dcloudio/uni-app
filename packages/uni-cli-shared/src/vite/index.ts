import fs from 'fs'
import path from 'path'
import type { Plugin } from 'vite'
import type { EmittedAsset } from 'rollup'
import type { ParserOptions } from '@vue/compiler-core'
import type {
  CompilerOptions,
  SFCStyleCompileOptions,
  TemplateCompiler,
} from '@vue/compiler-sfc'
import type { UniViteCopyPluginOptions } from './plugins/copy'
export { getIsStaticFile } from './plugins/vitejs/plugins/static'

export const cssTarget = 'chrome53'
export interface CopyOptions {
  /**
   * 静态资源，配置的目录，在 uni_modules 中同样支持
   */
  assets?: string[]
  targets?: UniViteCopyPluginOptions['targets']
}

interface UniVitePluginUniOptions {
  compiler?: TemplateCompiler
  styleOptions?: Pick<SFCStyleCompileOptions, 'postcssPlugins'>
  compilerOptions?: {
    miniProgram?: {
      emitFile?: (emittedFile: EmittedAsset) => string
    }
    isNativeTag?: ParserOptions['isNativeTag']
    isCustomElement?: ParserOptions['isCustomElement']
    directiveTransforms?: CompilerOptions['directiveTransforms']
    nodeTransforms?: CompilerOptions['nodeTransforms']
    whitespace?: CompilerOptions['whitespace']
  }
  jsxOptions?: {
    babelPlugins?: any[]
  }
  copyOptions?: CopyOptions | (() => CopyOptions)
}
export interface UniVitePlugin extends Plugin {
  uni?: UniVitePluginUniOptions
}

export * from './utils'
export * from './plugins'
export * from './features'
export * from './autoImport'

export * from './cloud'
export * from './extApi'

// https://github.com/vitejs/vite/blob/aac2ef77521f66ddd908f9d97020b8df532148cf/packages/vite/src/node/server/searchRoot.ts#L38
// vite 在初始化阶段会执行 initTSConfck，此时会 searchForWorkspaceRoot，如果找到了 pnpm-workspace.yaml 文件，会将其作为 root
// HBuilderX 项目，root 一定是 UNI_INPUT_DIR，所以需要重写 fs.existsSync，不重写的话，可能会找错，
// 一旦找错目录，而该目录下有 N 多文件目录，会导致遍历及其缓慢
export function rewriteExistsSyncHasRootFile() {
  const existsSync = fs.existsSync
  const pnpmWorkspaceYaml = path.join(
    process.env.UNI_INPUT_DIR,
    'pnpm-workspace.yaml'
  )
  fs.existsSync = (path) => {
    if (path === pnpmWorkspaceYaml) {
      return true
    }
    return existsSync(path)
  }
}
