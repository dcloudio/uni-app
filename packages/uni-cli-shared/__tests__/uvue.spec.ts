import { replaceScriptLang } from '../src/vite'

describe('uvue', () => {
  test('replaceScriptLang add lang="uts" if no lang attribute', () => {
    const code = `<template></template>
  <script>
  export default {}
  </script>
  <style>
  </style>
  `
    expect(replaceScriptLang('test.vue', code, false)).toMatchSnapshot()
  })
  test('replaceScriptLang replace lang="ts" to lang="uts"', () => {
    const code = `<template></template>
  <script lang="ts">
  export default {}
  </script>
  `
    expect(replaceScriptLang('test.vue', code, false)).toMatchSnapshot()
  })
  test('replaceScriptLang with lang="uts" not change', () => {
    const code = `<template></template>
  <script lang="uts">
  export default {}
  </script>
  `
    expect(replaceScriptLang('test.vue', code, false)).toMatchSnapshot()
  })
})
