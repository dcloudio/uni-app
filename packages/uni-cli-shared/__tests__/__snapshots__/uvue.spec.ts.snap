// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uvue replaceScriptLang add lang="uts" if no lang attribute 1`] = `
{
  "code": "<template></template>
  <script lang="uts">
  export default {}
  </script>
  <style>
  </style>
  ",
  "map": null,
}
`;

exports[`uvue replaceScriptLang replace lang="ts" to lang="uts" 1`] = `
{
  "code": "<template></template>
  <script lang="uts">
  export default {}
  </script>
  ",
  "map": null,
}
`;

exports[`uvue replaceScriptLang with lang="uts" not change 1`] = `undefined`;
